Add crawl errors endpoint

tw4l opened the pull request:
Fixes #738 

If this endpoint is called while the crawl is running, errors are pulled directly from redis.

If this endpoint is called when the crawl is finished, errors are pulled from mongodb, where they're written when crawls complete.

Displaying all errors while a crawl is running will require polling the endpoint until it completes, at which point all of the errors are available in the database.

The endpoint returns a 500 error with detail `redis_connection_error` if it's unable to connect to Redis.

NOTE: Test will fail until the crawler image with `--logErrorsToRedis` is pushed (see https://github.com/webrecorder/browsertrix-crawler/pull/279, now merged). With a crawler image built from that PR/branch, the new nightly test passes.

tw4l committed:
Add crawl errors endpoint

If this endpoint is called while the crawl is running, errors are
pulled directly from redis.

If this endpoint is called when the crawl is finished, errors are
pulled from mongodb, where they're written when crawls complete.

tw4l committed:
Add nightly backend test for errors endpoint

tw4l committed:
Add errors for failed and cancelled crawls to mongo

ikreymer commented on backend/test_nightly/conftest.py:
```diff
@@ -150,3 +150,33 @@ def crawl_config_info(admin_auth_headers, default_org_id):
+        "runNow": True,
+        "name": "Youtube crawl with errors",
+        "config": {
+            "seeds": [{"url": "https://www.youtube.com/watch?v=Sh-x3QmbRZc"}],
```
Hm, wonder if we should do this as part of nightly test, or something more 'predictable', such as list of URLs with some being invalid?

tw4l replied:
Invalid URLs don't seem to be throwing errors currently, so until we fix that I think this might be our best option?

ikreymer commented on backend/btrixcloud/crawls.py:
```diff
@@ -713,6 +717,17 @@ async def filter_crawl_queue(self, crawl_id, regex):
+            errors = await redis.lrange(f"{crawl_id}:e", 0, -1)
+        except exceptions.ConnectionError:
+            # pylint: disable=raise-missing-from
+            raise HTTPException(status_code=500, detail="redis_connection_error")
```
500 is usually unexpected crashes, maybe should do 503 or 400 here? I guess 503 is probably most accurate.

ikreymer commented on backend/btrixcloud/crawls.py:
```diff
@@ -713,6 +717,17 @@ async def filter_crawl_queue(self, crawl_id, regex):
+        """Get crawl errors from Redis and optionally store in mongodb."""
+        try:
+            redis = await self.get_redis(crawl_id)
+            errors = await redis.lrange(f"{crawl_id}:e", 0, -1)
```
Here we could probably fetch in batches, with offset and default size, so frontend can handle pagination, eg. `errors?offset=...`

tw4l replied:
makes sense! may as well use our existing pagination format for consistency's sake - I'll make that change

ikreymer commented on backend/btrixcloud/crawl_job.py:
```diff
@@ -235,9 +237,18 @@ async def finish_crawl(self):
 
+    async def add_crawl_errors_to_mongo(self):
+        """Pull crawl errors from redis and write to mongo"""
+        errors = await self.redis.lrange(f"{self.job_id}:e", 0, -1)
```
We might also want to do this incrementally, depending on how many errors we'll have or limit the error list size, eg. if this is pulling down, say a 10000+ list, this may likely crash the redis pod, given current memory limit.

Or, could just store first N entries.
Could do essentially what we're doing here:
https://github.com/webrecorder/browsertrix-crawler/blob/main/util/state.js#L272

tw4l replied:
makes sense!

tw4l committed:
Make code review revisions

- Paginate crawl errors API endpoint
- Load errors into mongo from redis 100 at a time
- Use 503 status code for connection error

tw4l commented:
@ikreymer This is ready for re-review. Tested the nightly test locally and manually ensured that errors are also available during crawling from Redis.

ikreymer committed:
Merge branch 'main' into issue-738-view-crawl-errors
